src/Features/Core/Portable/IntroduceUsingStatement/AbstractIntroduceUsingStatementCodeRefactoringProvider.cs

```diff
@@ -225,46 +224,91 @@ private static (SyntaxTriviaList sameLine, SyntaxTriviaList endOfLine) SplitTrai
                 }
 
-                if (node == declarationSyntax)
+                // Add new variables that were declared in the statement
+                var declaredVariables = semanticModel.GetAllDeclaredSymbols(node, cancellationToken);
+                foreach (var declaredVariable in declaredVariables)
                 {
-                    break; // Ignore the declaration and usages prior to the declaration
+                    variableDeclarationIndex[declaredVariable] = statementIndex;
+                    lastVariableUsageIndex[declaredVariable] = statementIndex;
+                    localVariables.Add(declaredVariable);
                 }
+            }
 
-                if (ContainsReference(node, localVariable, semanticModel, syntaxFactsService, cancellationToken))
+            var lastUsageIndex = 0;
+
+            // Walk through the local variables in the order that they were declared.
+            foreach (var localSymbol in localVariables)
+            {
+                var declarationIndex = variableDeclarationIndex[localSymbol];
+                if (declarationIndex > lastUsageIndex)
                 {
-                    return node;
+                    // If the variable was declared after the last statement to include in
+                    // the using statement, we have gone far enough and other variables will
+                    // also be declared outside the using statement.
+                    break;
                 }
+
+                lastUsageIndex = Math.Max(lastUsageIndex, lastVariableUsageIndex[localSymbol]);
             }
 
-            return null;
+            variableDeclarationIndex.Free();
+            lastVariableUsageIndex.Free();
+
+            return statementsAfterDeclaration[lastUsageIndex];
         }
 
-        private static bool ContainsReference(
+        private static ImmutableArray<ISymbol> GetReferencedLocalVariables(
             SyntaxNode node,
-            ILocalSymbol localVariable,
+            List<ISymbol> localVariables,
             SemanticModel semanticModel,
             ISyntaxFactsService syntaxFactsService,
             CancellationToken cancellationToken)
         {
+            var referencedVariables = ArrayBuilder<ISymbol>.GetInstance();
             if (syntaxFactsService.IsIdentifierName(node))
             {
                 var identifierName = syntaxFactsService.GetIdentifierOfSimpleName(node).ValueText;
 
-                return syntaxFactsService.StringComparer.Equals(localVariable.Name, identifierName) &&
-                    localVariable.Equals(semanticModel.GetSymbolInfo(node).Symbol);
+                var variable = localVariables.FirstOrDefault(localVariable
+                    => syntaxFactsService.StringComparer.Equals(localVariable.Name, identifierName) &&
+                        localVariable.Equals(semanticModel.GetSymbolInfo(node).Symbol));
+
+                referencedVariables.AddIfNotNull(variable);
```

Comment:
feels like this should return a set not an array, since you can add the same variable multiple times, and that's nto helpful.